### PR TITLE
[HW] Add hw.wire operation

### PR DESCRIPTION
**Pull request by fabianschuiki**
Add the `hw.wire` operation to the HW dialect. A wire offers the ability to assign a human-readable name or an inner symbol to an edge in the SSA graph, for human consumption or remote references through a symbol.

This fills a representational gap where language frontends like FIRRTL have to resort to `sv.wire` if they need any of the naming or symbol reference capabilities of a wire, but in doing so they also have to buy into the SV-specific quirks and connect semantics of `sv.wire`, which are annoying to analyze. Almost all uses in FIRRTL don't care about the connect semantics though. The new `hw.wire` op allows for expressing a wire without the connect semantics, leveraging the graph-ness of the HW dialect, which makes analysis a lot easier and will allow frontends to significantly reduce the number of SV-specific ops they emit. Decoupling frontends from the SV emission backend is very valuable in practice, as it allows for the intermediate HW representation to be consumed by other tools like simulators without them having to try and implement the SV semantics of a wire.

In a nutshell, this allows for the following replacement in almost all cases:

```mlir
// before
%myWire = sv.wire sym @mySym : !hw.inout<i42>
%0 = sv.read_inout %myWire : !hw.inout<i42>
sv.assign %myWire, %1 : i42
hw.output %0 : i42

// after
%myWire = hw.wire %1 sym @mySym : i42
hw.output %myWire : i42
```

This commit adds the `hw.wire` op, canonicalizations, and tests for it. It also adds `BuilderUtils` with a helper to accept `StringAttr`, `StringRef`, or `Twine` in a single builder argument, which is going to be useful for other operations as well. The changes to ExportVerilog are limited to qualifying `WireOp` as `sv::WireOp` to disambiguate from the new `hw::WireOp`.

### CODE REVIEW

File: include/circt/Dialect/HW/HWMiscOps.td [hunk 1]
```diff
@@ -53,6 +54,52 @@ def ConstantOp
...
+    SameOperandsAndResultType,
+    DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>]> {
+  let summary = "Assign a name or symbol to an SSA edge";
+  let description = [{
+    An `hw.wire` is used to assign a human-readable name or a symbol for remote
+    references to an SSA edge. It takes a single operand and returns that
+    its value unchanged as a result. The operation guarantees the following:
+
+    - The value of its operand remains observable under the given human-readable
+      name in any final output, or under the given symbol within the IR.
```
Comment (mikeurbach) [inline include/circt/Dialect/HW/HWMiscOps.td:67]: Maybe add something like "if specified" to be clear that name and inner_sym are optional, and this behavior is only present if one or the other is specified? Also, "remains observable" may be too strong... it looks like the canonicalizer will try to propagate the name attribute as a sv.namehint, but this is only a hint, and doesn't guarantee that the name will be observed.

**Review by mikeurbach — APPROVED** [review]
Some nitty comments about the specifics of the name/inner_sym and observability of names, but in general looks good to me.

File: include/circt/Dialect/HW/HWMiscOps.td (hunk 1 above)
Comment (fabianschuiki) [reply include/circt/Dialect/HW/HWMiscOps.td:67]: Yeah great point! I'll reword this to match exactly how the canonicalizer works right now. At some point we might want to specifically indicate on the op if a name is droppable or *must* be preserved. The way it is right now preserves how `sv.wire` behaves, which seems like a good starting point.

File: include/circt/Dialect/HW/HWMiscOps.td [hunk 2]
```diff
@@ -53,6 +54,52 @@ def ConstantOp
...
+    - The users of its result will always observe the operand through the
+      operation itself, meaning that optimizations cannot bypass the wire. This
+      ensures that if the wire's value is *forced*, for example through a
+      Verilog force statement, the forced value will affect all users of the
+      wire in the output.
```
Comment (youngar) [inline include/circt/Dialect/HW/HWMiscOps.td:72]: Since you have folders for the wire, I think that these are only guaranteed if it has a symbol?

Comment (youngar) [reply include/circt/Dialect/HW/HWMiscOps.td:72]: I see Mike beat me to it 😄

**Review by youngar — APPROVED** [review]

**Comment by fabianschuiki** [comment]
Updated the description to reflect how the canonicalizer treats the wire. Maybe in the future we'll want names to be non-droppable my default, and have something like "droppable_name" on the wire 🤔

File: include/circt/Support/BuilderUtils.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+#include "llvm/ADT/TypeSwitch.h"
+
+namespace circt {
+
+/// A helper union that can represent a `StringAttr`, `StringRef`, or `Twine`.
+/// It is intended to be used as arguments to an op's `build` function. This
+/// allows a single builder to accept any flavor value for a string attribute.
+/// The `get` function can then be used to obtain a `StringAttr` from any of the
+/// possible variants `StringAttrOrRef` can take.
+class StringAttrOrRef {
```
Comment (nandor) [inline include/circt/Support/BuilderUtils.h:22]: This is a Pandora's box. Can't we get away with two constructors for wire instead? One to take two StringAttrs and one to take two Twines and re-write the instantiation sites to eliminate mixed uses?

Comment (fabianschuiki) [reply include/circt/Support/BuilderUtils.h:22]: The problem is that we use this `StringRef` and `StringAttr` duality all over the place, and there seems to be a definite desire to have builders that accept `StringRef`. But always having the full canonical list of combinations seems like an anti-pattern. Ideally we'd be able to upstream some of these helpers to MLIR eventually, also for arrays and friends, such that TableGen can directly generate them.

**Review by teqdruid — APPROVED** [review]
Looks great! Will this replace `hw.probe`? The description of `hw.probe` makes it seem like there's a lot of overlap.

File: include/circt/Dialect/HW/HWMiscOps.td [hunk 4]
```diff
@@ -53,6 +54,66 @@ def ConstantOp
...
+      persists until code generation the resulting wire will have this name,
+      with a potential suffix to ensure uniqueness. If the wire is canonicalized
+      away, its name is propagated to its input operand as a name hint.
+
+    - The users of its result will always observe the operand through the
+      operation itself, meaning that optimizations cannot bypass the wire. This
+      ensures that if the wire's value is *forced*, for example through a
+      Verilog force statement, the forced value will affect all users of the
+      wire in the output.
+
```
Comment (uenoku) [inline include/circt/Dialect/HW/HWMiscOps.td:79]: Does it mean we can force the value of HWWireOp?

**Review by uenoku — APPROVED** [review]
LGTM!

File: include/circt/Dialect/HW/HWMiscOps.td (hunk 4 above)
Comment (fabianschuiki) [reply include/circt/Dialect/HW/HWMiscOps.td:79]: Yeah I think we want to define hw.wire such that forcing is possible, because that's one of the use cases we are seeing with sv.wire as well. And it creates a nice contract where hw.wire allows you to say "I want to later have access to this very SSA graph edge, and if I somehow forcefully change its value in unexpected ways, I want all ops downstream of it to see that change reflect." (Which implies that IMCP for example can't push constants through.)

Comment (darthscsi) [reply include/circt/Dialect/HW/HWMiscOps.td:79]: Optional debug port ala firrtl memories is my leaning.
` %myWire, %myWire.ref = hw.wire %1 : i32, ref<i32>`
